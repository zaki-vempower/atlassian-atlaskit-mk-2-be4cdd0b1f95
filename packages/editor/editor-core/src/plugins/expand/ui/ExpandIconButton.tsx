import React, { useCallback } from 'react';
import Button from '@atlaskit/button';
import {
  akEditorSwoopCubicBezier,
  expandMessages,
  ExpandLayoutWrapper,
} from '@atlaskit/editor-common';
import ChevronRightIcon from '@atlaskit/icon/glyph/chevron-right';
import { colors } from '@atlaskit/theme';
import Tooltip from '@atlaskit/tooltip';
import { InjectedIntl } from 'react-intl';
import { expandClassNames } from './class-names';

interface ExpandIconButtonProps {
  allowInteractiveExpand: boolean;
  expanded: boolean;
  intl?: InjectedIntl;
}

interface ExpandIconButtonWithLabelProps extends ExpandIconButtonProps {
  label: string;
}

const withTooltip = (WrapperComponent: React.ElementType) => {
  return class WithSortableColumn extends React.Component<
    ExpandIconButtonWithLabelProps
  > {
    constructor(props: ExpandIconButtonWithLabelProps) {
      super(props);
    }

    render() {
      const { label } = this.props;

      return (
        <Tooltip content={label} position="top" tag={ExpandLayoutWrapper}>
          <WrapperComponent {...this.props} />
        </Tooltip>
      );
    }
  };
};

const CustomButton = (props: ExpandIconButtonWithLabelProps) => {
  const { label, allowInteractiveExpand } = props;
  const useTheme = useCallback(
    (currentTheme: any, themeProps: any) => {
      const { buttonStyles, ...rest } = currentTheme(themeProps);
      return {
        buttonStyles: {
          ...buttonStyles,
          height: '100%',
          '& svg': {
            transform: props.expanded
              ? 'transform: rotate(90deg);'
              : 'tranform: rotate(0deg);',
            transition: `transform 0.2s ${akEditorSwoopCubicBezier};`,
          },
        },
        ...rest,
      };
    },
    [props],
  );

  return (
    <Button
      appearance="subtle"
      className={expandClassNames.iconContainer}
      iconBefore={<ChevronRightIcon label={label} primaryColor={colors.N80A} />}
      shouldFitContainer
      theme={useTheme}
      isDisabled={!allowInteractiveExpand}
    ></Button>
  );
};

const ButtonWithTooltip = withTooltip(CustomButton);
const ButtonWithoutTooltip = CustomButton;

export const ExpandIconButton = (props: ExpandIconButtonProps) => {
  const { expanded, intl } = props;
  const message = expanded
    ? expandMessages.collapseNode
    : expandMessages.expandNode;
  const label = (intl && intl.formatMessage(message)) || message.defaultMessage;

  if (props.allowInteractiveExpand) {
    return <ButtonWithTooltip label={label} {...props} />;
  }

  return (
    <ExpandLayoutWrapper>
      <ButtonWithoutTooltip label={label} {...props} />
    </ExpandLayoutWrapper>
  );
};
