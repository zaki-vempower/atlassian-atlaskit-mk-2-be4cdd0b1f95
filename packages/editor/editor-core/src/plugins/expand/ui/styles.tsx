import { css } from 'styled-components';
import { colors, gridSize } from '@atlaskit/theme';
import {
  blockNodesVerticalMargin,
  akLayoutGutterOffset,
  sharedExpandStyles,
} from '@atlaskit/editor-common';

import { expandClassNames } from './class-names';
import { BreakoutCssClassName } from '../../breakout/constants';

const EXPAND_SELECTED_BACKGROUND = 'rgba(255, 255, 255, 0.6)';

export const expandStyles = css`
  .${expandClassNames.icon} > div {
    display: flex;
  }

  .${expandClassNames.prefix} {
    ${sharedExpandStyles.ContainerStyles}
    margin-top: ${blockNodesVerticalMargin}rem;
    overflow: hidden;
    cursor: pointer;

    td > & {
      margin-top: 0;
    }

    .${expandClassNames.iconContainer} svg {
      transform: rotate(90deg);
    }
  }

  .ProseMirror > .${expandClassNames.type('expand')}, .${
  BreakoutCssClassName.BREAKOUT_MARK_DOM
} > .${expandClassNames.type('expand')} {
    margin-left: -${akLayoutGutterOffset}px;
    margin-right: -${akLayoutGutterOffset}px;
  }

  .${expandClassNames.content} {
    ${sharedExpandStyles.ContentStyles}
    cursor: text;
    padding-top: 0px;
  }

  .${expandClassNames.titleInput} {
    ${sharedExpandStyles.TitleInputStyles}
  }

  .${expandClassNames.titleContainer} {
    ${sharedExpandStyles.TitleContainerStyles};
    overflow: visible;
  }

  .${expandClassNames.expanded} {
    background: ${EXPAND_SELECTED_BACKGROUND};
    border-color: ${colors.N40A};

    .${expandClassNames.content} {
      padding-top: ${gridSize}px;
    }
  }

  .${expandClassNames.inputContainer} {
    width: 100%;
  }

  .${expandClassNames.prefix}:not(.${expandClassNames.expanded}) {
    .ak-editor-expand__content {
      position: absolute;
      height: 1px;
      width: 1px;
      overflow: hidden;
      clip: rect(1px, 1px, 1px, 1px);
      white-space: nowrap;
    }

    .${expandClassNames.iconContainer} svg {
      transform: rotate(0deg);
    }

    &:not(.ProseMirror-selectednode):not(.danger) {
      background: transparent;
      border-color: transparent;

      &:hover {
        border-color: ${colors.N50A};
        background: ${EXPAND_SELECTED_BACKGROUND};
      }
    }
  }
`;
