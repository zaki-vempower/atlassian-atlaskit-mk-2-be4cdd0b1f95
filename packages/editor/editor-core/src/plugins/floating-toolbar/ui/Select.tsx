import * as React from 'react';
import { Component, HTMLAttributes, ComponentClass } from 'react';
import styled from 'styled-components';

import Select, { ValueType } from '@atlaskit/select';

export interface RenderOptionsPropsT<T> {
  hide: () => void;
  dispatchCommand: (command: T) => void;
}

export interface SelectOption {
  value: string;
  label: string;
  selected?: boolean;
  disabled?: boolean;
  hidden?: boolean;
}

export interface Props {
  hideExpandIcon?: boolean;
  options: SelectOption[];
  dispatchCommand: (command: Function) => void;
  mountPoint?: HTMLElement;
  boundariesElement?: HTMLElement;
  scrollableElement?: HTMLElement;
  defaultValue?: SelectOption;
  placeholder?: string;
  onChange?: (change: ValueType<SelectOption>) => void;
  width?: number;
}

export interface State {
  isOpen: boolean;
}

const SelectWrapper: ComponentClass<HTMLAttributes<{}> & {
  width: number;
}> = styled.div`
  width: ${props => props.width}px;
`;

export default class Search extends Component<Props, State> {
  state: State = { isOpen: false };
  render() {
    const {
      options,
      onChange,
      defaultValue,
      placeholder,
      width = 200,
    } = this.props;
    return (
      <SelectWrapper width={width}>
        <Select<SelectOption>
          options={options}
          value={defaultValue}
          onChange={onChange}
          placeholder={placeholder}
          spacing={'compact'}
          menuPlacement="auto"
        />
      </SelectWrapper>
    );
  }
}
