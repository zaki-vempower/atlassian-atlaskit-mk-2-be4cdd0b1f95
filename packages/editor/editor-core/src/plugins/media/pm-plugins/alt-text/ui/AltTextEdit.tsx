import React from 'react';
import { KeyboardEvent } from 'react';
import { EditorView } from 'prosemirror-view';
import styled from 'styled-components';
import { colors } from '@atlaskit/theme';
import ChevronLeftLargeIcon from '@atlaskit/icon/glyph/chevron-left-large';
import CrossCircleIcon from '@atlaskit/icon/glyph/cross-circle';
import { messages } from '../messages';
import { injectIntl, InjectedIntlProps } from 'react-intl';
import Button from '../../../../floating-toolbar/ui/Button';
import PanelTextInput from '../../../../../ui/PanelTextInput';
import * as keymaps from '../../../../../keymaps';
import { closeMediaAltTextMenu, updateAltText } from '../commands';
import {
  withAnalyticsEvents,
  WithAnalyticsEventsProps,
} from '@atlaskit/analytics-next';
import {
  ACTION_SUBJECT,
  ACTION_SUBJECT_ID,
  EVENT_TYPE,
  fireAnalyticsEvent,
  ACTION,
  MediaAltTextActionType,
  FireAnalyticsCallback,
} from '../../../../analytics';

export const CONTAINER_WIDTH_IN_PX = 350;
const SupportText = styled.p`
  color: ${colors.N100};
  font-size: 12px;
  padding-right: 24px;
  padding-left: 28px;
  padding-top: 12px;
  padding-bottom: 12px;
  line-height: 20px;
  border-top: 1px solid ${colors.N30};
  margin: 0;
`;

const Container = styled.div`
  width: ${CONTAINER_WIDTH_IN_PX}px;
  display: flex;
  flex-direction: column;
  overflow: auto;
  line-height: 2;
`;

const InputWrapper = styled.section`
  display: flex;
  line-height: 0;
  padding: 0;
  align-items: center;
  margin-bottom: 4px;
`;

const BackButtonWrapper = styled.div`
  display: flex;
  margin-right: 4px;
`;

const ClearText = styled.span`
  color: ${colors.N80};
`;

type Props = {
  view: EditorView;
  value?: string;
} & InjectedIntlProps &
  WithAnalyticsEventsProps;

export type AltTextEditComponentState = {
  showClearTextButton: boolean;
};

export class AltTextEditComponent extends React.Component<
  Props,
  AltTextEditComponentState
> {
  private fireCustomAnalytics?: FireAnalyticsCallback;
  state = {
    showClearTextButton: Boolean(this.props.value),
  };

  constructor(props: Props) {
    super(props);

    const { createAnalyticsEvent } = props;
    this.fireCustomAnalytics = fireAnalyticsEvent(createAnalyticsEvent);
  }

  prevValue: string | undefined;

  componentDidMount() {
    this.prevValue = this.props.value;
  }

  componentWillUnmount() {
    this.fireAnalytics(ACTION.CLOSED);
    if (!this.prevValue && this.props.value) {
      this.fireAnalytics(ACTION.ADDED);
    }
    if (this.prevValue && !this.props.value) {
      this.fireAnalytics(ACTION.CLEARED);
    }
    if (this.prevValue && this.prevValue !== this.props.value) {
      this.fireAnalytics(ACTION.EDITED);
    }
  }

  render() {
    const {
      intl: { formatMessage },
      value,
    } = this.props;
    const { showClearTextButton } = this.state;

    const backButtonMessage = formatMessage(messages.back);
    const backButtonMessageComponent = keymaps.renderTooltipContent(
      backButtonMessage,
      keymaps.escape,
      'Esc',
    );

    return (
      <Container>
        <InputWrapper>
          <BackButtonWrapper>
            <Button
              title={formatMessage(messages.back)}
              icon={
                <ChevronLeftLargeIcon label={formatMessage(messages.back)} />
              }
              tooltipContent={backButtonMessageComponent}
              onClick={this.closeMediaAltTextMenu}
            />
          </BackButtonWrapper>
          <PanelTextInput
            placeholder={formatMessage(messages.placeholder)}
            defaultValue={value ? value : ''}
            onCancel={this.dispatchCancelEvent}
            onChange={this.handleOnChange}
            onSubmit={this.closeMediaAltTextMenu}
            autoFocus
          />
          {showClearTextButton && (
            <Button
              title={formatMessage(messages.clear)}
              icon={
                <ClearText>
                  <CrossCircleIcon label={formatMessage(messages.clear)} />
                </ClearText>
              }
              tooltipContent={formatMessage(messages.clear)}
              onClick={this.handleClearText}
            />
          )}
        </InputWrapper>
        <SupportText>{formatMessage(messages.supportText)}</SupportText>
      </Container>
    );
  }

  private closeMediaAltTextMenu = () => {
    const { view } = this.props;
    closeMediaAltTextMenu(view.state, view.dispatch);
  };

  private fireAnalytics(actionType: MediaAltTextActionType) {
    const { createAnalyticsEvent } = this.props;
    if (createAnalyticsEvent && this.fireCustomAnalytics) {
      this.fireCustomAnalytics({
        payload: {
          action: actionType,
          actionSubject: ACTION_SUBJECT.MEDIA,
          actionSubjectId: ACTION_SUBJECT_ID.ALT_TEXT,
          eventType: EVENT_TYPE.TRACK,
        },
      });
    }
  }

  private dispatchCancelEvent = (event: KeyboardEvent) => {
    const { view } = this.props;

    // We need to pass down the ESCAPE keymap
    // because when we focus on the Toolbar, Prosemirror blur,
    // making all keyboard shortcuts not working
    view.someProp('handleKeyDown', (fn: any) => fn(view, event));
  };

  private updateAltText = (newAltText: string) => {
    const { view } = this.props;
    const newValue = newAltText.length === 0 ? null : newAltText;
    updateAltText(newValue)(view.state, view.dispatch);
  };

  private handleOnChange = (newAltText: string) => {
    this.setState({
      showClearTextButton: Boolean(newAltText),
    });
    this.updateAltText(newAltText);
  };

  private handleClearText = () => {
    this.handleOnChange('');
  };
}

export default withAnalyticsEvents()(injectIntl(AltTextEditComponent));
