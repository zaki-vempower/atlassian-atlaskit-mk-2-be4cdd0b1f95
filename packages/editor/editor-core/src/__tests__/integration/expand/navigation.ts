import { BrowserTestCase } from '@atlaskit/webdriver-runner/runner';
import { editable, getDocFromElement, fullpage } from '../_helpers';
import emptyExpandAdf from './__fixtures__/empty-expand.json';
import twoLineExpandAdf from './__fixtures__/two-line-expand.json';
import doubleExpand from './__fixtures__/double-expand.json';

import {
  goToEditorTestingExample,
  mountEditor,
} from '../../__helpers/testing-example-helpers';

import { expandClassNames } from '../../../plugins/expand/ui/class-names';

const expandContentSelector = `.${expandClassNames.content} p`;

const focusExpandTitle = async (page: any) => {
  await page.browser.execute((selector: string) => {
    const input = document.querySelector('.' + selector);
    if (input) {
      // @ts-ignore
      input.focus();
    }
  }, expandClassNames.titleInput);
};

const collapseExpandThenFocusTitle = async (page: any) => {
  await page.click(`.${expandClassNames.icon}`);
  await focusExpandTitle(page);
};

BrowserTestCase(
  'navigation.ts: pressing Backspace should delete an expand when cursor is inside content',
  { skip: ['ie', 'safari', 'firefox'] },
  async (client: any, testName: string) => {
    const page = await goToEditorTestingExample(client);

    await mountEditor(page, {
      appearance: fullpage.appearance,
      defaultValue: emptyExpandAdf,
      UNSAFE_allowExpand: true,
    });

    await page.click(expandContentSelector);
    await page.keys('Backspace');

    const doc = await page.$eval(editable, getDocFromElement);
    expect(doc).toMatchCustomDocSnapshot(testName);
  },
);

BrowserTestCase(
  'navigation.ts: pressing Backspace should delete an expand when cursor is inside title',
  { skip: ['ie', 'safari', 'firefox'] },
  async (client: any, testName: string) => {
    const page = await goToEditorTestingExample(client);

    await mountEditor(page, {
      appearance: fullpage.appearance,
      defaultValue: emptyExpandAdf,
      UNSAFE_allowExpand: true,
    });

    await collapseExpandThenFocusTitle(page);
    await page.keys('Backspace');

    const doc = await page.$eval(editable, getDocFromElement);
    expect(doc).toMatchCustomDocSnapshot(testName);
  },
);

BrowserTestCase(
  'navigation.ts: pressing Enter should collapse an expand when cursor is inside title',
  { skip: ['ie', 'safari', 'firefox'] },
  async (client: any, testName: string) => {
    const page = await goToEditorTestingExample(client);

    await mountEditor(page, {
      appearance: fullpage.appearance,
      defaultValue: emptyExpandAdf,
      UNSAFE_allowExpand: true,
    });

    await collapseExpandThenFocusTitle(page);

    const doc = await page.$eval(editable, getDocFromElement);
    expect(doc).toMatchCustomDocSnapshot(testName);
  },
);

BrowserTestCase(
  'navigation.ts: pressing ArrowDown should set cursor after collapsed expand when it is in a title',
  { skip: ['ie', 'safari', 'firefox'] },
  async (client: any, testName: string) => {
    const page = await goToEditorTestingExample(client);

    await mountEditor(page, {
      appearance: fullpage.appearance,
      defaultValue: emptyExpandAdf,
      UNSAFE_allowExpand: true,
    });

    await collapseExpandThenFocusTitle(page);
    await page.keys('ArrowDown');
    await page.type(editable, 'abc');

    const doc = await page.$eval(editable, getDocFromElement);
    expect(doc).toMatchCustomDocSnapshot(testName);
  },
);

BrowserTestCase(
  'navigation.ts: pressing ArrowDown should set cursor from expand title to the content',
  { skip: ['ie', 'safari', 'firefox'] },
  async (client: any, testName: string) => {
    const page = await goToEditorTestingExample(client);

    await mountEditor(page, {
      appearance: fullpage.appearance,
      defaultValue: emptyExpandAdf,
      UNSAFE_allowExpand: true,
    });

    await focusExpandTitle(page);
    await page.keys('ArrowDown');
    await page.type(editable, 'abc');

    const doc = await page.$eval(editable, getDocFromElement);
    expect(doc).toMatchCustomDocSnapshot(testName);
  },
);

BrowserTestCase(
  'navigation.ts: pressing ArrowRight from an expand title should set the right GapCursor',
  { skip: ['ie', 'safari', 'firefox'] },
  async (client: any, testName: string) => {
    const page = await goToEditorTestingExample(client);

    await mountEditor(page, {
      appearance: fullpage.appearance,
      defaultValue: emptyExpandAdf,
      UNSAFE_allowExpand: true,
    });

    await focusExpandTitle(page);
    await page.keys('ArrowRight');
    await page.type(editable, 'after');

    const doc = await page.$eval(editable, getDocFromElement);
    expect(doc).toMatchCustomDocSnapshot(testName);
  },
);

BrowserTestCase(
  'navigation.ts: pressing ArrowLeft from an expand title should set the left GapCursor',
  { skip: ['ie', 'safari', 'firefox'] },
  async (client: any, testName: string) => {
    const page = await goToEditorTestingExample(client);

    await mountEditor(page, {
      appearance: fullpage.appearance,
      defaultValue: emptyExpandAdf,
      UNSAFE_allowExpand: true,
    });

    await focusExpandTitle(page);
    await page.keys('ArrowLeft');
    await page.type(editable, 'before');

    const doc = await page.$eval(editable, getDocFromElement);
    expect(doc).toMatchCustomDocSnapshot(testName);
  },
);

BrowserTestCase(
  'navigation.ts: pressing ArrowUp from an expand title should set the left GapCursor',
  { skip: ['ie', 'safari', 'firefox'] },
  async (client: any, testName: string) => {
    const page = await goToEditorTestingExample(client);

    await mountEditor(page, {
      appearance: fullpage.appearance,
      defaultValue: emptyExpandAdf,
      UNSAFE_allowExpand: true,
    });

    await focusExpandTitle(page);
    await page.keys('ArrowUp');
    await page.type(editable, 'before');

    const doc = await page.$eval(editable, getDocFromElement);
    expect(doc).toMatchCustomDocSnapshot(testName);
  },
);

BrowserTestCase(
  'navigation.ts: pressing ArrowUp from an expand content should focus the title',
  { skip: ['ie', 'safari', 'firefox'] },
  async (client: any, testName: string) => {
    const page = await goToEditorTestingExample(client);

    await mountEditor(page, {
      appearance: fullpage.appearance,
      defaultValue: emptyExpandAdf,
      UNSAFE_allowExpand: true,
    });

    await page.click(expandContentSelector);
    await page.type(editable, 'content');
    await page.keys('ArrowUp');
    await page.type(editable, 'title');

    const doc = await page.$eval(editable, getDocFromElement);
    expect(doc).toMatchCustomDocSnapshot(testName);
  },
);

BrowserTestCase(
  'navigation.ts: pressing ArrowUp from a second line of an expand content should not focus the title',
  { skip: ['ie', 'safari', 'firefox'] },
  async (client: any, testName: string) => {
    const page = await goToEditorTestingExample(client);

    await mountEditor(page, {
      appearance: fullpage.appearance,
      defaultValue: twoLineExpandAdf,
      UNSAFE_allowExpand: true,
    });

    await page.click(expandContentSelector);
    await page.keys('ArrowDown');
    await page.keys('ArrowUp');
    await page.type(editable, 'test');

    const doc = await page.$eval(editable, getDocFromElement);
    expect(doc).toMatchCustomDocSnapshot(testName);
  },
);

BrowserTestCase(
  'navigation.ts: pressing ArrowDown from the left GapCursor position should focus the title when an expand is collapsed',
  { skip: ['ie', 'safari', 'firefox'] },
  async (client: any, testName: string) => {
    const page = await goToEditorTestingExample(client);

    await mountEditor(page, {
      appearance: fullpage.appearance,
      defaultValue: emptyExpandAdf,
      UNSAFE_allowExpand: true,
    });

    await collapseExpandThenFocusTitle(page);
    await page.keys('ArrowLeft');
    await page.keys('ArrowDown');
    await page.type(editable, 'title');

    const doc = await page.$eval(editable, getDocFromElement);
    expect(doc).toMatchCustomDocSnapshot(testName);
  },
);

BrowserTestCase(
  'navigation.ts: pressing ArrowUp from the right GapCursor position should focus the title when an expand is collapsed',
  { skip: ['ie', 'safari', 'firefox'] },
  async (client: any, testName: string) => {
    const page = await goToEditorTestingExample(client);

    await mountEditor(page, {
      appearance: fullpage.appearance,
      defaultValue: emptyExpandAdf,
      UNSAFE_allowExpand: true,
    });

    await collapseExpandThenFocusTitle(page);
    await page.keys('ArrowRight');
    await page.keys('ArrowUp');
    await page.type(editable, 'title');

    const doc = await page.$eval(editable, getDocFromElement);
    expect(doc).toMatchCustomDocSnapshot(testName);
  },
);

describe('when there is a expanded followed by another', () => {
  describe('when both are collapsed', () => {
    describe('and the focus are inside of the first title', () => {
      BrowserTestCase(
        'pressing ArrowDown should create a gap cursor on the left',
        { skip: ['ie'] },
        async (client: any, testName: string) => {
          const page = await goToEditorTestingExample(client);

          await mountEditor(page, {
            appearance: fullpage.appearance,
            defaultValue: doubleExpand,
            UNSAFE_allowExpand: true,
          });

          await page.click(
            '.ak-editor-expand__title-input[value="First title"]',
          );
          await page.keys('ArrowDown');
          await page.keys('I am here'.split(''));

          const doc = await page.$eval(editable, getDocFromElement);
          expect(doc).toMatchCustomDocSnapshot(testName);
        },
      );
    });
  });
});

BrowserTestCase(
  'navigation.ts: when cursor is after a collapsed expand, pressing Backspace should focus the title',
  { skip: ['ie', 'safari', 'firefox'] },
  async (client: any, testName: string) => {
    const page = await goToEditorTestingExample(client);

    await mountEditor(page, {
      appearance: fullpage.appearance,
      defaultValue: emptyExpandAdf,
      UNSAFE_allowExpand: true,
    });

    await collapseExpandThenFocusTitle(page);
    await page.keys('ArrowDown');
    await page.keys('Backspace');
    await page.type(editable, 'title');

    const doc = await page.$eval(editable, getDocFromElement);
    expect(doc).toMatchCustomDocSnapshot(testName);
  },
);

BrowserTestCase(
  'navigation.ts: when cursor is after two collapsed expands, pressing Backspace should focus the title of the first one',
  { skip: ['ie', 'safari', 'firefox'] },
  async (client: any, testName: string) => {
    const page = await goToEditorTestingExample(client);

    await mountEditor(page, {
      appearance: fullpage.appearance,
      defaultValue: doubleExpand,
      UNSAFE_allowExpand: true,
    });

    await page.click('.ak-editor-expand__title-input[value="Second title"]');
    await page.keys('ArrowDown');
    await page.keys('Backspace');
    await page.type(editable, 'hello');

    const doc = await page.$eval(editable, getDocFromElement);
    expect(doc).toMatchCustomDocSnapshot(testName);
  },
);
