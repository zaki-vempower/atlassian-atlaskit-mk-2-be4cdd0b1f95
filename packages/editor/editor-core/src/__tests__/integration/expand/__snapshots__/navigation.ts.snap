// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`navigation.ts: pressing ArrowDown from the left GapCursor position should focus the title when an expand is collapsed 1`] = `
Object {
  "content": Array [
    Object {
      "attrs": Object {
        "__expanded": false,
        "title": "title",
      },
      "content": Array [
        Object {
          "type": "paragraph",
        },
      ],
      "type": "expand",
    },
    Object {
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`navigation.ts: pressing ArrowDown should set cursor after collapsed expand when it is in a title 1`] = `
Object {
  "content": Array [
    Object {
      "attrs": Object {
        "__expanded": false,
        "title": "",
      },
      "content": Array [
        Object {
          "type": "paragraph",
        },
      ],
      "type": "expand",
    },
    Object {
      "content": Array [
        Object {
          "text": "abc",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`navigation.ts: pressing ArrowDown should set cursor from expand title to the content 1`] = `
Object {
  "content": Array [
    Object {
      "attrs": Object {
        "__expanded": true,
        "title": "",
      },
      "content": Array [
        Object {
          "content": Array [
            Object {
              "text": "abc",
              "type": "text",
            },
          ],
          "type": "paragraph",
        },
      ],
      "type": "expand",
    },
    Object {
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`navigation.ts: pressing ArrowLeft from an expand title should set the left GapCursor 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": "before",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "attrs": Object {
        "__expanded": true,
        "title": "",
      },
      "content": Array [
        Object {
          "type": "paragraph",
        },
      ],
      "type": "expand",
    },
    Object {
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`navigation.ts: pressing ArrowRight from an expand title should set the right GapCursor 1`] = `
Object {
  "content": Array [
    Object {
      "attrs": Object {
        "__expanded": true,
        "title": "",
      },
      "content": Array [
        Object {
          "type": "paragraph",
        },
      ],
      "type": "expand",
    },
    Object {
      "content": Array [
        Object {
          "text": "after",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`navigation.ts: pressing ArrowUp from a second line of an expand content should not focus the title 1`] = `
Object {
  "content": Array [
    Object {
      "attrs": Object {
        "__expanded": true,
        "title": "",
      },
      "content": Array [
        Object {
          "content": Array [
            Object {
              "text": "onetest",
              "type": "text",
            },
          ],
          "type": "paragraph",
        },
        Object {
          "content": Array [
            Object {
              "text": "two",
              "type": "text",
            },
          ],
          "type": "paragraph",
        },
      ],
      "type": "expand",
    },
  ],
  "type": "doc",
}
`;

exports[`navigation.ts: pressing ArrowUp from an expand content should focus the title 1`] = `
Object {
  "content": Array [
    Object {
      "attrs": Object {
        "__expanded": true,
        "title": "title",
      },
      "content": Array [
        Object {
          "content": Array [
            Object {
              "text": "content",
              "type": "text",
            },
          ],
          "type": "paragraph",
        },
      ],
      "type": "expand",
    },
    Object {
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`navigation.ts: pressing ArrowUp from an expand title should set the left GapCursor 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": "before",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "attrs": Object {
        "__expanded": true,
        "title": "",
      },
      "content": Array [
        Object {
          "type": "paragraph",
        },
      ],
      "type": "expand",
    },
    Object {
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`navigation.ts: pressing ArrowUp from the right GapCursor position should focus the title when an expand is collapsed 1`] = `
Object {
  "content": Array [
    Object {
      "attrs": Object {
        "__expanded": false,
        "title": "title",
      },
      "content": Array [
        Object {
          "type": "paragraph",
        },
      ],
      "type": "expand",
    },
    Object {
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`navigation.ts: pressing Backspace should delete an expand when cursor is inside content 1`] = `
Object {
  "content": Array [
    Object {
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`navigation.ts: pressing Backspace should delete an expand when cursor is inside title 1`] = `
Object {
  "content": Array [
    Object {
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`navigation.ts: pressing Enter should collapse an expand when cursor is inside title 1`] = `
Object {
  "content": Array [
    Object {
      "attrs": Object {
        "__expanded": false,
        "title": "",
      },
      "content": Array [
        Object {
          "type": "paragraph",
        },
      ],
      "type": "expand",
    },
    Object {
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`navigation.ts: when cursor is after a collapsed expand, pressing Backspace should focus the title 1`] = `
Object {
  "content": Array [
    Object {
      "attrs": Object {
        "__expanded": false,
        "title": "title",
      },
      "content": Array [
        Object {
          "type": "paragraph",
        },
      ],
      "type": "expand",
    },
    Object {
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`navigation.ts: when cursor is after two collapsed expands, pressing Backspace should focus the title of the first one 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aenean mi nisl, venenatis eget auctor vitae, venenatis quis lorem. Suspendisse maximus tortor vel dui tincidunt cursus. Vestibulum magna nibh, auctor non auctor id, finibus vitae orci. Nulla viverra ipsum et nunc fringilla ultricies. Pellentesque vitae felis molestie justo finibus accumsan. Suspendisse potenti. Nulla facilisi. Integer dignissim quis velit quis elementum. Sed sit amet varius ante. Duis vestibulum porta augue eu laoreet. Morbi id risus et augue sollicitudin aliquam. In et ligula dolor. Nam ac aliquet diam.",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "attrs": Object {
        "__expanded": false,
        "title": "First title",
      },
      "content": Array [
        Object {
          "type": "paragraph",
        },
      ],
      "type": "expand",
    },
    Object {
      "attrs": Object {
        "__expanded": false,
        "title": "Second titlehello",
      },
      "content": Array [
        Object {
          "type": "paragraph",
        },
      ],
      "type": "expand",
    },
    Object {
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`pressing ArrowDown should create a gap cursor on the left 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aenean mi nisl, venenatis eget auctor vitae, venenatis quis lorem. Suspendisse maximus tortor vel dui tincidunt cursus. Vestibulum magna nibh, auctor non auctor id, finibus vitae orci. Nulla viverra ipsum et nunc fringilla ultricies. Pellentesque vitae felis molestie justo finibus accumsan. Suspendisse potenti. Nulla facilisi. Integer dignissim quis velit quis elementum. Sed sit amet varius ante. Duis vestibulum porta augue eu laoreet. Morbi id risus et augue sollicitudin aliquam. In et ligula dolor. Nam ac aliquet diam.",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "attrs": Object {
        "__expanded": false,
        "title": "First title",
      },
      "content": Array [
        Object {
          "type": "paragraph",
        },
      ],
      "type": "expand",
    },
    Object {
      "content": Array [
        Object {
          "text": "I am here",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "attrs": Object {
        "__expanded": false,
        "title": "Second title",
      },
      "content": Array [
        Object {
          "type": "paragraph",
        },
      ],
      "type": "expand",
    },
    Object {
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;
